// astro/dist/assets/fonts/metrics.js
import { fontFamilyToCamelCase } from "@capsizecss/metrics";
import { fromBuffer } from "@capsizecss/unpack";
const QUOTES_RE = /^["']|["']$/g;
const withoutQuotes = (str) => str.trim().replace(QUOTES_RE, "");
const metricCache = {};
function filterRequiredMetrics({
  ascent,
  descent,
  lineGap,
  unitsPerEm,
  xWidthAvg
}) {
  return {
    ascent,
    descent,
    lineGap,
    unitsPerEm,
    xWidthAvg
  };
}
async function getMetricsForFamily(family) {
  family = withoutQuotes(family);
  if (family in metricCache) return metricCache[family];
  try {
    const name = fontFamilyToCamelCase(family);
    const { entireMetricsCollection } = await import("@capsizecss/metrics/entireMetricsCollection");
    const metrics = entireMetricsCollection[name];
    if (!("descent" in metrics)) {
      metricCache[family] = null;
      return null;
    }
    const filteredMetrics = filterRequiredMetrics(metrics);
    metricCache[family] = filteredMetrics;
    return filteredMetrics;
  } catch {
    metricCache[family] = null;
    return null;
  }
}
async function readMetrics(family, buffer) {
  const metrics = await fromBuffer(buffer);
  metricCache[family] = filterRequiredMetrics(metrics);
  return metricCache[family];
}
function toPercentage(value, fractionDigits = 4) {
  const percentage = value * 100;
  return `${+percentage.toFixed(fractionDigits)}%`;
}
function toCSS(properties, indent = 2) {
  return Object.entries(properties).map(([key, value]) => `${" ".repeat(indent)}${key}: ${value};`).join("\n");
}
function generateFallbackFontFace(metrics, fallback) {
  const {
    name: fallbackName,
    font: fallbackFontName,
    metrics: fallbackMetrics,
    ...properties
  } = fallback;
  const preferredFontXAvgRatio = metrics.xWidthAvg / metrics.unitsPerEm;
  const fallbackFontXAvgRatio = fallbackMetrics ? fallbackMetrics.xWidthAvg / fallbackMetrics.unitsPerEm : 1;
  const sizeAdjust = fallbackMetrics && preferredFontXAvgRatio && fallbackFontXAvgRatio ? preferredFontXAvgRatio / fallbackFontXAvgRatio : 1;
  const adjustedEmSquare = metrics.unitsPerEm * sizeAdjust;
  const ascentOverride = metrics.ascent / adjustedEmSquare;
  const descentOverride = Math.abs(metrics.descent) / adjustedEmSquare;
  const lineGapOverride = metrics.lineGap / adjustedEmSquare;
  const declaration = {
    "font-family": JSON.stringify(fallbackName),
    src: `local(${JSON.stringify(fallbackFontName)})`,
    "size-adjust": toPercentage(sizeAdjust),
    "ascent-override": toPercentage(ascentOverride),
    "descent-override": toPercentage(descentOverride),
    "line-gap-override": toPercentage(lineGapOverride),
    ...properties
  };
  return `@font-face {
${toCSS(declaration)}
}
`;
}
export {
  generateFallbackFontFace,
  getMetricsForFamily,
  readMetrics
};
